keps/sig-api-machinery/3028-retry-during-server-initialization/README.md

```diff
@@ -0,0 +1,221 @@
+```
+StatusCode = 503
+
+Header:
+ Retry-After: N
+ X-Kubernetes-Ready: false
+```
+
+- A status code of `503` with a header `Retry-After: N` in the response instructs the client to
+  retry the request after `N` seconds.
+- The response header `X-Kubernetes-Ready: false` indicates to the caller that the server understood 
+  the `X-Kubernetes-If-Ready` request header and took action appropriately.
+  
+If the server has fully initialized, the request will be processed and the following header will be injected to the response.
+
+```
+Header:
+ X-Kubernetes-Ready: true
+```
+
+Upon receipt of a `503` status code with `Retry-After` response header, the K8s client-go logic will 
+automatically retry the request up to maximum retry threshold.
+
+The caller will receive the following result:
+- After certain retries the request has succeeded, and the caller will see
+  `X-Kubernetes-Ready: true` in the response header.
+- The underlying client-go retry logic will give up after maximum retry threshold has been reached and will 
+  propagate the following error to the caller:
+```
+StatusCode = 503
+
+Header:
+ Retry-After: N
+ X-Kubernetes-Ready: false
+```
+
+With this, we can establish the following invariants:
+- Presence of the header in the response indicates that the server understands the `X-Kubernetes-If-Ready` 
+  header and has taken action appropriately.
+- On the other hand, absence of the header in the response indicates that the server does not
+  understand the `X-Kubernetes-If-Ready` header
+
+This gives the client enough information to decide what to do in a situation where it sends `X-Kubernetes-If-Ready` 
+to the server that does not understand the header due to a version mismatch.
+
+#### client-go Changes
+
+- Add a new boolean field `IfReadyHeader` in `rest.Config`. This allows a component author to
+  opt-in for this feature for all requests associated with a `rest.Config` instance.
+
+- Use a custom `http.RoundTripper` that adds the header to the request inside `RoundTrip`
+
+#### No Automatic Retry Approach
+
+Another option for the server is to always attach the header in the response:
+```
+Header:
+ X-Kubernetes-Ready: {true|false}
+```
+
+when it sees the `X-Kubernetes-If-Ready` header in the request, and the request is processed as usual. The caller
+checks the response status code and the value of the `X-Kubernetes-Ready` header and then react accordingly. With this 
+approach, the retry rests with the caller since client-go will not automatically retry here.
+
+
+#### Server Run Option
+- We propose adding a new boolean server option `startup-send-retry-after-if-not-initialized`. The cluster operator can set 
+  this option to `true` to enable this feature on the server side. By default, this option will be disabled.
+
+Given that the conditionality is opt-in, what is the motivation for adding this server option? If the server logic
```

Comment:
I think a featuregate is enough.  Once this is GA, I'd like to have consistent apiserver behavior with this always on.